### ISSUE TITLE
Generic conditional type narrowing doesn't cascade as expected

### ISSUE THREAD

**Issue by milohansen** [microsoft/TypeScript#46975]
# Bug Report

I've tried my best to search for related issues and read the documentation thoroughly but failed to find any useful information. I'm sorry if this bug turns out to be duplicate or even not a bug at all.

### 🔎 Search Terms

conditional type if undefined

### 🕗 Version & Regression Information

All versions available in the playground (http://www.typescriptlang.org/play)
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about Generics and Type Guards.

### ⏯ Playground Link

[Playground link](https://www.typescriptlang.org/play?ts=4.5.2#code/C4TwDgpgBAYg9nAPAFShAHsCA7AJgZymwFcBbAIwgCcoAfKfYKgS2wHMA+KAXigG8AsACgoUAG4BDADbEIALijIA3MIC+w4aEhQAchKpU4AdxRpMOAkTKUa9Ri3Z0oxPBABmrCLi69UGLHiELrjunrhQAPxEEGLUUArwSMgcSkA)

### 💻 Code

```ts
type Foo<T extends number | string> = {
  value: T;
}

type Narrow<T extends number | string | undefined> = T extends undefined ? never : Foo<T>;
//                                                                                     ^
// ERROR: Type 'string | number | undefined' is not assignable to type 'string | number'.
```

### 🙁 Actual behavior

In the above example, `T` throws an error when passed to `Foo` because the compiler believes `T` could be undefined. 

### 🙂 Expected behavior

The condition has checked if `T` extends undefined so I would expect `T`'s type to be narrowed from `string | number | undefined` to `string | number` on the right side of the ternary operator.

**Comment by RyanCavanaugh** [comment]
I feel like there might be a theoretical justification for why we don't do this, but I can't reconstruct it at the moment

**Comment by jack-williams** [comment]
This is related, but I'm not sure if it is out-of-date: https://github.com/microsoft/TypeScript/issues/29188#issuecomment-451238804

FWIW, this works
```typescript
type Narrow<T extends number | string | undefined> = T extends number | string  ? Foo<T> : never;
```

**Comment by titouandk** [comment]
I’ve run into a very similar issue with conditional types, `ReturnType`, and unions. Here’s a distilled example:

```ts
export type EndFunction = (...args: any[]) => any;

export type SyncFactoryFunction = (...args: any[]) => EndFunction;
export type AsyncFactoryFunction = (...args: any[]) => Promise<EndFunction>;
export type FactoryFunction = SyncFactoryFunction | AsyncFactoryFunction;

export type EndFunctionParameters<F extends FactoryFunction> =
  F extends SyncFactoryFunction
    ? Parameters<ReturnType<F>> // TS raises an error here!
    : F extends AsyncFactoryFunction
      ? Parameters<Awaited<ReturnType<F>>>
      : never;
```

TypeScript raises:

```
Type 'ReturnType<F>' does not satisfy the constraint '(...args: any) => any'.
Type 'EndFunction | Promise<EndFunction>' is not assignable to type '(...args: any) => any'.
```

### Why this happens
The problem is that `ReturnType<F>` evaluates to a **union** (`EndFunction | Promise<EndFunction>`) *before* the conditional type (`F extends ...`) has a chance to narrow it. Since `Promise<EndFunction>` does not extend `(...args: any) => any`, the constraint check fails, even though in the branch where `F` is `SyncFactoryFunction`, the type is safe.

In other words, conditional types don’t “wait” to distribute over `ReturnType` before validating constraints.

### Workaround

In my case, I have to avoid using `ReturnType` here, and instead use inference inside the conditional branch:

```ts
type EndFunctionParameters<F extends FactoryFunction> =
  F extends (...args: any[]) => infer R
    ? R extends (...args: infer P) => any
      ? P
      : R extends Promise<(...args: infer P) => any>
        ? P
        : never
    : never;
```

This way, inference happens branch by branch, and `Promise<...>` is handled explicitly.

**Comment by RyanCavanaugh** [comment]
This is a duplicate of [#29188](https://github.com/microsoft/TypeScript/issues/29188). Conditional types track constraints in the true branch, but the false branch would need a negated type to represent `T` excluding `undefined`. Reversing the test, as in `T extends number | string ? Foo<T> : never`, puts the useful constraint in the true branch.
